modules/dxp/apps/search-experiences/search-experiences-web/src/main/resources/META-INF/resources/sxp_blueprint_admin/js/edit_sxp_blueprint/query_builder_tab/SelectTypes.js

```diff
@@ -5,39 +5,183 @@
 
 import ClayButton from '@clayui/button';
 import ClayIcon from '@clayui/icon';
-import ClayTable from '@clayui/table';
-import React, {useContext} from 'react';
+import ClayList from '@clayui/list';
+import React, {useContext, useState} from 'react';
 
 import ThemeContext from '../../shared/ThemeContext';
 import SearchableTypesModal from './SearchableTypesModal';
+import SelectSubtypes from './SelectSubtypes';
+
+/**
+ * Grabs all the selected subtypes of a certain type.
+ *
+ * @param {array} selected
+ * @param {string} className
+ * @returns {array}
+ */
+const getSelectedSubtypes = (selected, className) => {
+	return selected.find((item) => item.type === className)?.subtypes;
+};
+
+/**
+ * Turns a flat list of types and subtypes into a nested array of the types
+ * with its subtypes as children. This is called when setting up the 'selected'
+ * state and allows for easier manipulation of the data.
+ *
+ * @param {*} initialSelectedTypes
+ * @param {*} subtypeClassesMap
+ * @returns {array}
+ */
+const setupSelected = (initialSelectedTypes, subtypeClassesMap = {}) => {
```

Comment:
I think this is fine as-is, but I wanted to point out being cautious about doing data transformations within child components. Anytime we're mutating data like this, it can lead to headaches in the future since now you'd have 2 different data representations of the same value.

A better approach would be immediately transforming this when you fetch the data and do all of your "transform to how backend expects the format" in one place directly where the form submission function happens.

Reply:
Noted! Yea, it would be nice to keep track of the data in one representation. It was just that the data looked so different between what would be submitted and how it was being used for display 😅 ... maybe we can address this in a later refactor!